src/NumericalAlgorithms/RootFinding/RootBracketing.hpp

```diff
@@ -0,0 +1,317 @@
+ * `f_at_lower_bound` and `f_at_upper_bound` are replaced with
+ * `f` evaluated at those bracketing points.
+ *
+ */
+template <typename Functor>
+void bracket_possibly_undefined_function_in_interval(
+    const gsl::not_null<double*> lower_bound,
+    const gsl::not_null<double*> upper_bound,
+    const gsl::not_null<double*> f_at_lower_bound,
+    const gsl::not_null<double*> f_at_upper_bound, const Functor& f,
+    const double guess) noexcept {
+  // Initial values of x1,x2,y1,y2.  Use `guess` and `upper_bound`,
+  // because in typical usage `guess` underestimates the actual
+  // root, and `lower_bound` is more likely than `upper_bound` to be
+  // invalid.
+  double x1 = guess;
+  double x2 = *upper_bound;
+  auto y1 = f(x1);
+  auto y2 = f(x2);
+  if (not(y1 and y2 and y1.get() * y2.get() <= 0.0)) {
+    // Root is not bracketed, so try to do so.
+    // First try lower_bound.
+    const double x3 = *lower_bound;
+    const auto y3 = f(x3);
+    if (y1 and y3 and y1.get() * y3.get() <= 0.0) {
+      // Bracketed! Throw out x2,y2.  Rename variables to keep x1 < x2.
+      x2 = x1;
+      y2 = y1;
+      x1 = x3;
+      y1 = y3;
+    } else if (y2 and y3 and y2.get() * y3.get() <= 0.0) {
+      // Bracketed! Throw out x1,y1.
+      x1 = x3;
+      y1 = y3;
+    } else {
+      // Our simple checks didn't work, so call the more general method.
+      std::array<double, 4> tmp = bracketing_detail::bracket_by_contracting(
+          {{x3, x1, x2}}, {{y3, y1, y2}}, f);
+      x1 = tmp[0];
+      x2 = tmp[1];
+      y1 = tmp[2];
+      y2 = tmp[3];
+    }
+  }
+  *f_at_lower_bound = y1.get();
+  *f_at_upper_bound = y2.get();
+  *lower_bound = x1;
+  *upper_bound = x2;
+}
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief Brackets the single root of the
+ * function `f` for each element in a `DataVector`, assuming the root
+ * lies in the given interval and that `f` may be undefined at some
+ * points in the interval.
+ *
+ * `f` is a binary invokable that takes a `double` and a `size_t` as
+ * arguments.  The `double` is the current value at which to evaluate
+ * `f`, and the `size_t` is the index into the `DataVector`s.  `f`
+ * returns a `boost::optional<double>` which evaluates to false if the
+ * function is undefined at the supplied point.
+ *
+ * Assumes that there is only one root in the interval.
+ *
+ * Assumes that if \f$f(x_1)\f$ and \f$f(x_2)\f$ are both defined for
+ * some \f$(x_1,x_2)\f$, then \f$f(x)\f$ is defined for all \f$x\f$
+ * between \f$x_1\f$ and \f$x_2\f$.
+ *
+ * On input, assumes that the root lies in the interval
+ * [`lower_bound`,`upper_bound`].  Optionally takes a `guess` for the
+ * location of the root.
+ *
+ * On return, `lower_bound` and `upper_bound` are replaced with values that
+ * bracket the root and for which the function is defined, and
+ * `f_at_lower_bound` and `f_at_upper_bound` are replaced with
+ * `f` evaluated at those bracketing points.
+ *
```

Comment:
document the erroring conditions: the range of sign change is smaller than 1/2^depth of the interval, or no root exists, and any other error conditions that would be valuable to know but I've overlooked.